dali/operators/random/choice.h

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_CHOICE_H_
+#define DALI_OPERATORS_RANDOM_CHOICE_H_
+
+#include <cstdint>
+#include <random>
+#include "dali/core/error_handling.h"
+#include "dali/operators/random/rng_base.h"
+#include "dali/operators/random/rng_base_cpu.h"
+#include "dali/operators/random/rng_base_gpu.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operator/arg_helper.h"
+
+#define DALI_CHOICE_0D_TYPES \
+  uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t
+#define DALI_CHOICE_1D_TYPES                                                                      \
+  bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float16, float, \
+      double
+
+namespace dali {
+
+/**
+ * @brief Variant for 1D input - we sample index and take the corresponding value from input pointer
+ */
+template <typename T, bool uniform, bool indirect = true>
+struct ChoiceSampleDist {
+  using DistType =
+      std::conditional_t<uniform, std::uniform_int_distribution<>, std::discrete_distribution<>>;
+
+  DALI_HOST_DEV explicit ChoiceSampleDist() {}
+
+  DALI_HOST_DEV ChoiceSampleDist(const T *elements, const float *p_first, const float *p_last)
+      : elements_(elements) {
+    if constexpr (!uniform) {
+      dist_ = DistType(p_first, p_last);
+    } else {
+      assert(false);  // Should not be called
+    }
+  }
+
+  DALI_HOST_DEV ChoiceSampleDist(const T *elements, int64_t element_count) : elements_(elements) {
+    if constexpr (uniform) {
+      dist_ = DistType(0, element_count - 1);
+    } else {
+      assert(false);  // Should not be called
+    }
+  }
+
+  template <typename Generator>
+  DALI_HOST_DEV T Generate(Generator &st) {
+    auto choice_idx = dist_(st);
+    return elements_[choice_idx];
+  }
+
+  const T *elements_ = nullptr;
+  DistType dist_ = {};
+};
+
+/**
+ * @brief Variant for generating indices. In case of 0D input, we treat those indices as the
```

Comment:
Nitpick: _Variant for generating indices_ is misleading for me. In this specialization we *don't* generate indices, but we generate actual values directly.

Reply:
done